### PR TITLE
Issue50 storage information

### PR DESCRIPTION
**Pull request by mortenwh**
**Summary**: Added storage_information

**Related issue**: #50 

**Suggested reviewer(s)**: @TAlonglong 

**Reviewer checklist**:

- [ ] The headers of all files contain a reference to the repository license (i.e., "License: This file is part of py-mmd-tools, licensed under the Apache License 2.0 (https://www.apache.org/licenses/LICENSE-2.0)")
- [ ] 100% test coverage of new code - meaning:
    - [ ] The overall test coverage increased or remained the same as before
    - [ ] Every function is accompanied with a test suite
    - [ ] Tests are both positive (testing that the function work as intended with valid data) and negative (testing that the function behaves as expected with invalid data, e.g., that correct exceptions are thrown)
    - [ ] Functions with optional arguments have separate tests for all options
- [ ] Examples are supported by doctests
- [ ] All tests are passing
- [ ] All names (e.g., files, classes, functions, variables) are explicit
- [ ] Documentation (as docstrings) is complete and understandable

The checklist is based on the S-ENDA conventions and definition of done (see https://s-enda-documentation.readthedocs.io/en/latest/general_conventions.html). The above points are not necessarily relevant to all contributions. In that case, please add a short explanation to help the reviewer.

### CODE REVIEW

**Comment by mortenwh** [comment]
I'm not sure if the path will be correct - maybe we have to override it? 

Can you check the case for thredds storage? I guess local files are only for testing?

File: py_mmd_tools/nc_to_mmd.py [hunk 1]
```diff
@@ -620,12 +622,28 @@ def to_mmd(self, *args, **kwargs):
...
+        md5hasher = FileHash('md5')
+        fchecksum = md5hasher.hash_file(self.netcdf_product)
+        self.metadata['storage_information'] = {
+                'file_name': os.path.basename(self.netcdf_product),
+                'file_location': os.path.abspath(self.netcdf_product).replace(os.path.basename(self.netcdf_product),''),
+                'file_format': 'NetCDF-CF',
+                'file_size': pathlib.Path(self.netcdf_product).stat().st_size / (1024 * 1024),
+                'file_size_unit': 'MB',
+                'checksum': fchecksum,
+                'checksum_type': '%ssum'%md5hasher.hash_algorithm,
```
Comment (TAlonglong) [inline py_mmd_tools/nc_to_mmd.py:636]: Why are you adding sum to the hash_algorithm type?

File: tests/test_nc_to_mmd.py [hunk 2]
```diff
@@ -536,6 +537,26 @@ def test_publication_date(self):
...
+
+    def test_checksum_is_equal_as_stored(self):
+        # NOTE: if the netcdf file is changed, this test will fail
+        fn = 'tests/data/reference_nc.nc'
+        nc2mmd = Nc_to_mmd(fn, check_only=True)
+        nc2mmd.to_mmd()
+        checksum = nc2mmd.metadata['storage_information']['checksum']
+        with open('tests/data/hashes.md5', 'r') as f:
+            ll=f.readline()
+        self.assertEqual(ll.split(' ')[0], checksum)
```
Comment (TAlonglong) [inline tests/test_nc_to_mmd.py:560]: I dont understand the point of this test. Is it to detect if the reference file has changed?

File: py_mmd_tools/nc_to_mmd.py (hunk 1 above)
Comment (mortenwh) [reply py_mmd_tools/nc_to_mmd.py:636]: that is needed as an attribute to the checksum field in MMD

**Comment by TAlonglong** [comment]
Looks like this small test fails @mortenwh 

>>> from filehash import FileHash
>>> md5hasher = FileHash('md5')
>>> md5hasher.hash_file('https://thredds.met.no/thredds/dodsC/remotesensingsatellite/polar-swath/2020/12/01/metopb-avhrr-20201201155244-20201201160030.nc')
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/home/trygveas/miniconda3/envs/mmd/lib/python3.8/site-packages/filehash/filehash.py", line 176, in hash_file
    with open(filename, mode="rb", buffering=0) as fp:
FileNotFoundError: [Errno 2] No such file or directory: 'https://thredds.met.no/thredds/dodsC/remotesensingsatellite/polar-swath/2020/12/01/metopb-avhrr-20201201155244-20201201160030.nc'

File: tests/test_nc_to_mmd.py (hunk 2 above)
Comment (mortenwh) [reply tests/test_nc_to_mmd.py:560]: yes - partly out of interest... It is probably not strictly necessary. Should I remove it?

**Comment by mortenwh** [comment]
Ok - then the file must probably probably be downloaded first... should we do that? It will probably require extra processing time..

**Comment by mortenwh** [comment]
I'm not too keen at testing online data streams, since they may fail for other reasons than errors in py-mmd-tools...

File: py_mmd_tools/nc_to_mmd.py (hunk 1 above)
Comment (TAlonglong) [reply py_mmd_tools/nc_to_mmd.py:636]: Yes I understand that. But here the `md5hasher.hash_algorithm` returns `md5`
And the `checksum_type` becomes `md5sum`. Why do you add `sum` to `md5` for the `checksum_type`?

Comment (mortenwh) [reply py_mmd_tools/nc_to_mmd.py:636]: that's required by MDD: "Checksums can be md5sum, shasum (sha256sum, sha384sum, sha512sum) and other higher order approaches."

**Comment by TAlonglong** [comment]
No, we should not download the file to get the md5.

**Comment by TAlonglong** [comment]
I guess you just add an exception `FileNotFoundError` and skip it.

File: tests/test_nc_to_mmd.py (hunk 2 above)
Comment (TAlonglong) [reply tests/test_nc_to_mmd.py:560]: I thin so

Comment (mortenwh) [reply tests/test_nc_to_mmd.py:560]: ok

**Comment by mortenwh** [comment]
I'll add downloading. It doesn't make sense to skip it, because then we will never have it given that we're going to harvest from thredds...

**Comment by TAlonglong** [comment]
Is it that important? Another way would be to let the nc_to_mmd ssh to a machine with lustre mounted and make the md5sum there. Or if it runs on k8s with lustre mounted you could generate it from the local file.

But its difficult to know the internal location. For satellite it is /lustre/storeB/immutable/archive/projects/remotesensing/satellite-thredds/

**Comment by mortenwh** [comment]
We could have an option to provide the local path and override the downloading?

**Comment by TAlonglong** [comment]
Yes, that was a good idea.

File: py_mmd_tools/nc_to_mmd.py [hunk 3]
```diff
@@ -615,11 +616,24 @@ def to_mmd(self, *args, **kwargs):
...
+                    if access['type'] == 'HTTP':
+                        resource = access['resource']
+                print('Downloading NetCDF file to calculate checksum...')
+                file_for_checksum_calculation = wget.download(resource)
+                rm_file_for_checksum_calculation = True
+            file_size = pathlib.Path(file_for_checksum_calculation).stat().st_size / (1024 * 1024)
         else:
             self.metadata['data_access'] = []
             file_for_checksum_calculation = self.netcdf_product
+            file_size = pathlib.Path(self.netcdf_product).stat().st_size / (1024 * 1024)
```
Comment (TAlonglong) [inline py_mmd_tools/nc_to_mmd.py:636]: file_size can be outside the if else using the file_for_checksum_calculation

File: py_mmd_tools/nc_to_mmd.py [hunk 4]
```diff
@@ -629,13 +643,16 @@ def to_mmd(self, *args, **kwargs):
         fchecksum = md5hasher.hash_file(file_for_checksum_calculation)
         self.metadata['storage_information'] = {
                 'file_name': os.path.basename(self.netcdf_product),
-                'file_location': os.path.abspath(self.netcdf_product).replace(os.path.basename(self.netcdf_product),''),
+                'file_location': self.netcdf_product.replace(os.path.basename(self.netcdf_product),''),
```
Comment (TAlonglong) [inline py_mmd_tools/nc_to_mmd.py:646]: Why are you using replace here?

Is it not like `os.path.dirname(self.netcdf_product)`

Comment (mortenwh) [reply py_mmd_tools/nc_to_mmd.py:646]: I think that will include the filename.. I tried a bit back and forth and concluded this worked

File: py_mmd_tools/nc_to_mmd.py (hunk 3 above)
Comment (mortenwh) [reply py_mmd_tools/nc_to_mmd.py]: yes - done now (without local testing..)

**Review by TAlonglong — APPROVED** [review]
LGTM

**Pull request merged**